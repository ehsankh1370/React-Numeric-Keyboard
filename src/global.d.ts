declare module '*.scss' {
  const content: Record<string, string>;
  export default content;
}
declare module '*.svg' {
  const content: string;
  export default content;
}
declare module '*.png' {
  const content: string;
  export default content;
}
