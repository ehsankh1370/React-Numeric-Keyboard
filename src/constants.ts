/**
 * fill array of numbers from 1 to 9
 */
export const numbers = Array.from({ length: 9 }, (_, i) => (i + 1).toString());

/**
 * Keyboard possible characters
 */
export const keyboardCharacters = {
  '0': '0',
  '1': '1',
  '2': '2',
  '3': '3',
  '4': '4',
  '5': '5',
  '6': '6',
  '7': '7',
  '8': '8',
  '9': '9',
  Backspace: 'Backspace',
};

/**
 * Default keyboard transition time
 */

export const DEFAULT_CLOSE_ANIMATION_TIME = 300;
