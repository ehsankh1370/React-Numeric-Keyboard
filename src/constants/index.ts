/**
 * fill array of numbers from 1 to 9
 */
const numbers = Array.from({ length: 9 }, (_, i) => (i + 1).toString());

/**
 * Keyboard possible characters
 */
const keyboardCharacters = {
  '0': '0',
  '1': '1',
  '2': '2',
  '3': '3',
  '4': '4',
  '5': '5',
  '6': '6',
  '7': '7',
  '8': '8',
  '9': '9',
  Backspace: 'Backspace',
};

/**
 * Default keyboard transition time
 */

const DEFAULT_CLOSE_ANIMATION_TIME = 300;

export { numbers, keyboardCharacters, DEFAULT_CLOSE_ANIMATION_TIME };
