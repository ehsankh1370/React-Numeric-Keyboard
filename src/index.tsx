export { NumericKeyboard } from './components/numeric-keyboard';
export type {KeyboardProps} from './types'
