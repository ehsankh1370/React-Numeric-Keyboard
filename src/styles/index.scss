@import './vars.scss';

:root {
  -webkit-tap-highlight-color: transparent;
}
