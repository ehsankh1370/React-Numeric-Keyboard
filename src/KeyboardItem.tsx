import React, { FC } from 'react';
//Constants
import { keyboardCharacters } from './constants';
//Utils
import { classNameGenerator } from './utils';
//Styles
import styles from './assets/index.module.scss';
//Types
import { MainProps } from '.';

type KeyboardItemProps = Pick<MainProps, 'backSpaceIcon' | 'theme'> & {
  item?: string;
  onClick?: (value: string | undefined) => void;
};

const KeyboardItem: FC<KeyboardItemProps> = (props): JSX.Element => {
  const { item, onClick, backSpaceIcon, children, theme } = props;

  const onKeyboardItemClick = (): void => {
    onClick?.(item);
  };

  const classNames = classNameGenerator([
    theme === 'light'
      ? styles.lightThemeKeyboardItem
      : styles.darkThemeKeyboardItem,
    styles.KeyboardItem,
  ]);

  return (
    <div className={classNames} onClick={onKeyboardItemClick}>
      {item ? (
        <>
          {item === keyboardCharacters.Backspace ? (
            backSpaceIcon || (
              <img
                src={
                  theme === 'light'
                    ? './assets/light-backspace.png'
                    : './assets/dark-backspace.png'
                }
                alt={keyboardCharacters.Backspace}
              />
            )
          ) : (
            <p className={styles.text}>{item}</p>
          )}
        </>
      ) : (
        children
      )}
    </div>
  );
};
export default KeyboardItem;
