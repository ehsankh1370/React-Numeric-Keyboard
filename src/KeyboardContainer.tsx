import React, { FC, useState } from 'react';
//Components
import KeyboardItem from './KeyboardItem';
//Constants
import { numbers, keyboardCharacters } from './constants';
//Utils
import { classNameGenerator } from './utils';
//Styles
import styles from './assets/index.module.scss';
//Types
import { KeyboardContainerProps } from './types';



const KeyboardContainer: FC<KeyboardContainerProps> = ({
  onChange,
  keyboardContainerClassName,
  isKeyboardDisabled,
  leftIcon,
  backSpaceIcon,
  theme,
}) => {
  const [value, setValue] = useState<string>('');

  const onKeyboardItemClick = (name: string): void => {
    if (!isKeyboardDisabled) {
      if (name === keyboardCharacters.Backspace) {
        onChange?.({ value: value.slice(0, -1), name });
        setValue(prevValue => prevValue.slice(0, -1));
      } else {
        onChange?.({ value: value + name, name });
        setValue(prevValue => prevValue + name);
      }
    }
  };
  const keyboardContainer = classNameGenerator([
    styles.itemsContainer,
    keyboardContainerClassName,
  ]);

  return (
    <div className={keyboardContainer}>
      {isKeyboardDisabled && <div className={styles.disabledLayout} />}
      {numbers.map(
        (item: string): JSX.Element => (
          <KeyboardItem
            key={item}
            item={item}
            onClick={(): void => onKeyboardItemClick(item)}
            theme={theme}
          />
        )
      )}

      {leftIcon ? (
        <KeyboardItem theme={theme}>{leftIcon}</KeyboardItem>
      ) : (
        <span className={styles.emptyField} />
      )}
      <KeyboardItem
        item={keyboardCharacters.zero}
        onClick={(): void => onKeyboardItemClick(keyboardCharacters.zero)}
        theme={theme}
      />
      <KeyboardItem
        item={keyboardCharacters.Backspace}
        onClick={(): void => onKeyboardItemClick(keyboardCharacters.Backspace)}
        backSpaceIcon={backSpaceIcon}
        theme={theme}
      />
    </div>
  );
};

export default KeyboardContainer;
