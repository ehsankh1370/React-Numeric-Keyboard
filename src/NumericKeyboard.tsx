import React, {
  FC,
  useState,
  useEffect,
  CSSProperties,
  RefObject,
} from 'react';
//Components
import KeyboardContainer from './KeyboardContainer';
//Utils
import { classNameGenerator } from './utils';
//styles
import styles from './assets/index.module.scss';

export interface MainProps {
  isOpen: boolean;
  hasInitialTransition?: boolean;
  transitionTime?: number;
  className?: string;
  isKeyboardDisabled?: boolean;
  onChange?: ({ value, name }: { value: string; name: string }) => void;
  backSpaceIcon?: React.ReactNode;
  leftIcon?: React.ReactNode;
  style?: CSSProperties;
  keyboardContainerClassName?: string;
  theme?: 'light' | 'dark';
  ref?: RefObject<HTMLDivElement>;
}

const DEFAULT_CLOSE_ANIMATION_TIME = 300;

const NumericKeyboard: FC<MainProps> = (props): JSX.Element | null => {
  const [isShow, setIsShow] = useState(true);
  const {
    isOpen,
    className,
    hasInitialTransition = true,
    transitionTime = DEFAULT_CLOSE_ANIMATION_TIME,
    onChange,
    isKeyboardDisabled = false,
    backSpaceIcon,
    leftIcon,
    style,
    keyboardContainerClassName,
    theme = 'light',
    ref,
  } = props;

  const animationClassesGenerator = (): string | boolean => {
    if (hasInitialTransition) {
      return isOpen ? styles.startAnimation : styles.closeAnimation;
    }
    return false;
  };
  const mainClassName = classNameGenerator([
    theme === 'light' ? styles.lightTheme : styles.darkTheme,
    styles.main,
    animationClassesGenerator(),
    className,
  ]);

  useEffect(() => {
    if (!isOpen && hasInitialTransition) {
      let timer = setTimeout(() => {
        setIsShow(false);
      }, transitionTime);
      return clearTimeout(timer);
    } else {
      setIsShow(isOpen);
    }
  }, [isOpen]);

  return isShow ? (
    <div
      className={mainClassName}
      style={{
        animationDuration: `${transitionTime / 1000}s`,
        ...style,
      }}
      ref={ref}
    >
      <KeyboardContainer
        onChange={onChange}
        backSpaceIcon={backSpaceIcon}
        isKeyboardDisabled={isKeyboardDisabled}
        leftIcon={leftIcon}
        keyboardContainerClassName={keyboardContainerClassName}
        theme={theme}
      />
    </div>
  ) : null;
};

export { NumericKeyboard };
