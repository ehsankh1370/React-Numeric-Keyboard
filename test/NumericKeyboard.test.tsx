import React from 'react';
import '@testing-library/jest-dom';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { NumericKeyboard } from '../src/NumericKeyboard';

describe('testing UI of the keyboard', () => {
  it('should render 1 to 9 numbers if keyboard is open', () => {
    const { getByText, queryByText } = render(<NumericKeyboard isOpen />);
    expect(getByText('5')).toBeDefined();

    expect(queryByText('10')).not.toBeInTheDocument();
  });
  it('should not render viewable keyboard item if isOpen props is false', () => {
    const { queryByText } = render(<NumericKeyboard isOpen={false} />);
    expect(queryByText('5')).toBeFalsy();
  });
  it('should accept custom class when provided', () => {
    const { container } = render(<NumericKeyboard isOpen className="custom" />);
    expect(container.firstChild).toHaveClass('custom');
  });
  it('should render custom backspace icon if provided', () => {
    const { queryByText } = render(
      <NumericKeyboard isOpen backSpaceIcon="my-custom-back" />
    );
    expect(queryByText('my-custom-back')).toBeInTheDocument();
  });
  it('should render custom left icon if provided', () => {
    const { queryByText } = render(
      <NumericKeyboard isOpen leftIcon="my-left-icon" />
    );
    expect(queryByText('my-left-icon')).toBeInTheDocument();
  });
});

describe('testing functionality of the keyboard', () => {
  it('should change value by clicking on the keyboard items', () => {
    const { getByText, getByAltText } = render(
      <NumericKeyboard isOpen onChange={onChange} />
    );
    fireEvent.click(getByText('1'));
    fireEvent.click(getByText('2'));
    fireEvent.click(getByText('5'));
    fireEvent.click(getByAltText('Backspace'));
    fireEvent.click(getByText('9'));
    function onChange({ value }: { value: string }) {
      waitFor(() => {
        expect(value).toBe('129');
      });
    }
  });
  it('keyboard should not be clickable if keyboard is disabled', () => {
    let values;
    const { getByText, rerender } = render(
      <NumericKeyboard isOpen onChange={onChange} isKeyboardDisabled={false} />
    );
    fireEvent.click(getByText('5'));
    fireEvent.click(getByText('4'));
    rerender(
      <NumericKeyboard isOpen onChange={onChange} isKeyboardDisabled={true} />
    );
    fireEvent.click(getByText('9'));
    function onChange({ value }: { value: string }) {
      values = value;
    }
    expect(values).toBe('54');
    expect(values).not.toBe('549');
  });
});

describe('snapshot test', () => {
  it('to match snapshot', () => {
    const { container } = render(<NumericKeyboard isOpen />);
    expect(container).toMatchSnapshot();
  });
});
