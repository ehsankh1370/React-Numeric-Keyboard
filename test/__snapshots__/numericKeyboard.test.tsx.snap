// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test to match snapshot 1`] = `
<div>
  <div
    class="main startAnimation"
    style="animation-duration: 0.3s;"
  >
    <div
      class="itemsContainer"
      data-keyboard-container="true"
    >
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          1
        </p>
      </div>
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          2
        </p>
      </div>
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          3
        </p>
      </div>
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          4
        </p>
      </div>
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          5
        </p>
      </div>
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          6
        </p>
      </div>
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          7
        </p>
      </div>
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          8
        </p>
      </div>
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          9
        </p>
      </div>
      <div
        class="KeyboardItem deactiveMode"
      />
      <div
        class="KeyboardItem isNumber"
      >
        <p
          class="text"
        >
          0
        </p>
      </div>
      <div
        class="KeyboardItem"
      >
        <img
          alt="Backspace"
          src=""
        />
      </div>
    </div>
  </div>
</div>
`;
